armory/scenarios/video_ucf101_scenario_adversarial_preloaded.py

```diff
@@ -0,0 +1,162 @@
+                    if defense_type == "Trainer":
+                        defense.fit(x, y, batch_size=batch_size, nb_epochs=1)
+                    else:
+                        classifier.fit(x, y, batch_size=batch_size, nb_epochs=1)
+
+        if defense_type == "Transform":
+            # NOTE: Transform currently not supported
+            logger.info(f"Transforming classifier with {defense_type} defense...")
+            defense = load_defense_wrapper(config["defense"], classifier)
+            classifier = defense()
+
+        classifier.set_learning_phase(False)
+
+        # Evaluate the ART classifier on benign test examples
+        logger.info(f"Loading test dataset {config['dataset']['name']}...")
+        test_data_generator = load_dataset(
+            config["dataset"],
+            epochs=1,
+            split_type="test",
+            preprocessing_fn=preprocessing_fn,
+        )
+
+        logger.info("Running inference on benign examples...")
+        metrics_logger = metrics.MetricsLogger.from_config(config["metric"])
+
+        for x_batch, y_batch in tqdm(test_data_generator, desc="Benign"):
+            for x, y in zip(x_batch, y_batch):
+                # combine predictions across all stacks
+                y_pred = np.mean(classifier.predict(x), axis=0)
+                metrics_logger.update_task(y, y_pred)
+        metrics_logger.log_task()
+
+        # Evaluate the ART classifier on adversarial test examples
+        logger.info("Loading / testing adversarial examples...")
+
+        if config["attack"]["preloaded"]:  # load existing adversarial dataset
+            test_data_generator = load_dataset(
+                config["attack"]["preloaded"],
+                epochs=1,
+                split_type="adversarial",
+                preprocessing_fn=preprocessing_fn,
+            )
+        else:  # generate attack using ART
+            attack = load_attack(config["attack"], classifier)
+            test_data_generator = load_dataset(
+                config["dataset"],
+                epochs=1,
+                split_type="test",
+                preprocessing_fn=preprocessing_fn,
+            )
+        for x_batch, y_batch in tqdm(test_data_generator, desc="Attack"):
+            if config["attack"]["preloaded"]:
+                attack_type = config["attack"]["preloaded"]["attack_type"]
+                x_batch = x_batch[attack_type]
+            for x, y in zip(x_batch, y_batch):
+                if config["attack"]["preloaded"]:
+                    x_adv = x
+                else:
+                    # each x is of shape (n_stack, 3, 16, 112, 112)
+                    #    n_stack varies
+                    attack.set_params(batch_size=x.shape[0])
+                    x_adv = attack.generate(x=x)
+
+                # combine predictions across all stacks
+                y_pred = np.mean(classifier.predict(x_adv), axis=0)
+                metrics_logger.update_task(y, y_pred, adversarial=True)
```

Comment:
This isn't pulling the pairs of (clean, adversarial), so when the perturbation is measured in line 141, it will be 0.

Reply:
I didn't see line 141.  Is it suppose to calculate some distance metric?  I can add in pulling of clean example before line 141.

Lines 146-162 is a template to pulling the pairs.

Reply:
Yes, 141 is calculating perturbation distance(s) between clean and adversarial for the list of metrics provided by the user. I think that having the default be to load both clean and adversarial and calculate those perturbations would work best here. I don't think that the overhead of loading both is going to be overly high.

Reply:
Got it.  I just saw that there's a "metric" key in the config.  Is your intention of "metric/perturbation" to match the distance used in generating the attack, or any user-defined distance metric?  If the latter, is there a way to calculate multiple metrics (L-0, L-2, L-inf, Wasserstein) in one pass, like passing a list to "metric/perturbation?"

Reply:
The intention is for any distance metric, though right now it only uses ones baked in to Armory, which is currently Lp (and L0) norms. You should be able to pass a list into the perturbation part of the metrics config and it will generate multiple values.